Fix Physics Material Assets in PhysXCommon to be processed in the AssetProcessor

spham-amzn opened the pull request:
## What does this PR do?
This fixes the situation where the Assets in the enabled 'PhysXCommon' gem was not being picked up by the Asset Processor as source assets. 

fixes https://github.com/o3de/o3de/issues/17700
fixes https://github.com/o3de/o3de/issues/17671 

## How was this PR tested?
On a new project with PhysX, PhysXCommon enabled, ran Asset Processor and confirmed the Physics Materials are processed.

nick-l-o3de commented:
Weird since physxcore depends on it
```
dependencies": [
        "PhysXCommon",
        ```
        but I guess this dependency is only enforced when adding gems via the gui project manager...

spham-amzn commented:
> Weird since physxcore depends on it (as in the PhysX gem), but I guess dependencies only add their dependent gems when you use the project manager GUI Scripts to alter a project, not if a template is wrong.
> 
> is it possible that other templates have PhysX but not PhysXcommon? Looks like the ScriptOnlyProject might also be missing it.
> 
> ```
> dependencies": [
>         "PhysXCommon",
> ```

This turned out to be a problem with how we do gems dependencies. When building, the dependent gem targets are added as a dependency when a build is triggered, however it turns out that phase where the aliases are built up, where logic like determining the builders alias to establish the setregs in builder targets does not consider dependent gems. That was the original reason why I added the `PhysXCommon` explicitly, but this is not a great solution. 

I've updated this PR to add a mechanism to perform a recursive dependency walking of gems to gather additional gem names so that the processing of the alias won't miss any dependent gems (that will be pulled into the build anyways)

lemonade-dm commented on cmake/Gems.cmake:
```diff
@@ -80,6 +80,72 @@ macro(o3de_gem_setup default_gem_name)
+
+endfunction()
+
+# resolve_all_dependent_gem_names
+#
+# Given a list of gem names, resolve all the dependent gems and build a list of additional
+# gems that need to be included for full dependency coverage
+#
+# \arg:input_gem_name_list(LIST) - The list of gem names to resolve all the dependencies
+# \arg:out_gem_name_list(LIST) - The list of additional gems that are not currently in the list to consider
+function(resolve_all_dependent_gem_names input_gem_name_list out_gem_name_list)
+
+    # Build up the list of all dependent gems (recursively)
+    unset(all_gems_name_list)
+    foreach(gem_name IN LISTS input_gem_name_list)
+        unset(dependent_gems_for_gem)
+        resolve_all_dependent_gem_names_recursive(${gem_name} "" dependent_gems_for_gem)
+        list(APPEND all_gems_name_list ${dependent_gems_for_gem})
+    endforeach()
+
+    # Remove duplicates and gem names that were already on the list
+    list(REMOVE_DUPLICATES all_gems_name_list)
+    foreach(gem_name IN LISTS input_gem_name_list)
+        list(REMOVE_ITEM all_gems_name_list ${gem_name})
+    endforeach()
+
+    list(SORT all_gems_name_list)
+    set(${out_gem_name_list} ${all_gems_name_list} PARENT_SCOPE)
```
This doesn't perform any caching of the gem_names that already had it's dependencies resolved in a cmake GLOBAL property.

However we already gather the list of Gem dependencies for each Gem based on the active Gem "variants".
The [o3de_add_variant_dependencies_for_gem_dependencies](https://github.com/o3de/o3de/blob/8749168e1f79263395b945f1623b7314003bf366/cmake/Gems.cmake#L134-L141) should be sufficient for this purpose without the need to add a  "resolve_all_dependent_gem_names" function.

However if we are going to implement such a function, it should be done in a way that the result per Gem name is cache in a cmake GLOBAL property. Because  `ly_enable_gems` is called multiple times through out the CMake configuration step this function can gather the dependent gem names multiple times.

Two suggestions here:
1. The list of dependent gem names can be gathering in the `ly_enable_gems_delayed` function at the end of this file. That function is only called once at the end of the CMake configuration step and iterates over all  Gems.
2. The caching of the "Gem Name" -> "dependent Gem Names"  map can be done in the `Subdirectories.cmake` [reorder_dependent_gems_with_cycle_detection](https://github.com/o3de/o3de/blob/development/cmake/Subdirectories.cmake#L275-L282) function as it already recursively reads the "dependents" field for each Gem and [caches](https://github.com/o3de/o3de/blob/development/cmake/Subdirectories.cmake#L320-L321) the paths into the "Dependent:${gem_external_subdir_path}" GLOBAL property.
In that same function a mapping since we have access to that mapping of "Gem external subdirectory" -> "Gem dependencies external subdirectories", the `o3de_read_json_key` function can be just be used to read the "gem_name" field for each and then a new mapping GLOBAL property of "Gem name" -> "Gem dependencies names" can be added there.

spham-amzn replied:
Thats a good suggestion, I can eliminate the additional loop and recursion if its already done somewhere (I didn't see that [reorder_dependent_gems_with_cycle_detection](https://github.com/o3de/o3de/blob/development/cmake/Subdirectories.cmake#L275-L282) function.

spham-amzn replied:
Updated the PR, with some notes:
1. The gem dependency resolution on names needed to occur before the [LY_DELAYED_ENABLE_GEMS](https://github.com/spham-amzn/o3de/blob/79232a97d2bbe37457c0b6c20bb10fac7d9001c5/cmake/Gems.cmake#L439) property for the project, `ly_enable_gems_delayed` was too late in the process
2. The reason was that the [LY_DELAYED_LOAD_](https://github.com/spham-amzn/o3de/blob/79232a97d2bbe37457c0b6c20bb10fac7d9001c5/cmake/Gems.cmake#L557-L558) property was set from iterating on the results of `LY_DELAYED_ENABLE_GEMS_"${project}"`. 
3. Setting the global property allowed me to memo-ize the recursion and simplify the logic

lemonade-dm replied:
That is good with me.

lemonade-dm commented on cmake/Gems.cmake:
```diff
@@ -87,62 +87,37 @@ endmacro()
-        resolve_all_dependent_gem_names_recursive(${gem_name} "" dependent_gems_for_gem)
-        list(APPEND all_gems_name_list ${dependent_gems_for_gem})
-    endforeach()
+        # Set the property for the gem so that we won't encounter a possible infinite recursion
+        # when evaluating the gem dependencies' dependency
+        set_property(GLOBAL PROPERTY GEM_DEPENDENCIES_"${gem_name}" ${gem_dependencies})
```
This seems kind of dangerous to partially set the gem name dependencies GLOBAL property part way through.
This would result in any gem name right before the start of a cycle receiving a partial list of the gem name dependencies, which could include itself.
ex. `Atom_RPI -> Atom_RHI -> Atom_RPI`.

Instead what could be done is to implement cycle detection logic by appending the visited gem names to a function scoped variable using `list(APPEND)`. 
If a cycle is detected, then a FATAL_ERROR can be logged as there would be no way to properly resolve the gem dependencies at that point.
